### PR TITLE
Add distributed prepartitioning check

### PR DESCRIPTION
**Pull request by matthiasdiener**

### CODE REVIEW

**Comment by matthiasdiener** [comment]
Is this going in the direction you had in mind @inducer ?

File: pytato/distributed.py [hunk 1]
```diff
@@ -1105,6 +1112,135 @@ class MissingRecvError(DistributedPartitionVerificationError):
...
+                            dest_rank=expr.send.dest_rank,
+                            comm_tag=expr.send.comm_tag,
+                            shape=expr.send.data.shape,
+                            dtype=expr.send.data.dtype)
+
+        new_name = self.name_generator()
+        self.output_name_to_send_node[new_name] = s
+
+    def map_distributed_send(self, expr: DistributedSend) -> None:
+        raise ValueError("Unpartitioned DAG should not have DistributedSend nodes")
```
Comment (inducer) [inline pytato/distributed.py:1149]: This isn't totally true; they're hidden in the ref holders. Rephrase.

File: pytato/distributed.py [hunk 2]
```diff
@@ -1105,6 +1112,135 @@ class MissingRecvError(DistributedPartitionVerificationError):
...
+        return id(expr)
+
+    def map_distributed_recv(  # type: ignore[override]
+            self, expr: DistributedRecv) -> None:
+        new_name = self.name_generator()
+        self.input_name_to_recv_node[new_name] = expr
+
+    def map_distributed_send_ref_holder(  # type: ignore[override]
+            self, expr: DistributedSendRefHolder) -> None:
+        s = _SummarizedDistributedSend(
```
Comment (inducer) [inline pytato/distributed.py:1138]: This is missing a recursion.

File: pytato/distributed.py [hunk 3]
```diff
@@ -1105,6 +1112,135 @@ class MissingRecvError(DistributedPartitionVerificationError):
     pass
 
 
+@optimize_mapper(drop_args=True, drop_kwargs=True, inline_get_cache_key=True)
+class _DistributedDAGGatherer(CachedWalkMapper):
```
Comment (inducer) [inline pytato/distributed.py:1116]: This does not find local data-flow edges. Maybe generalize off of `UsersCollector`, [as in the old partitioner](https://github.com/inducer/pytato/blob/baf7f5fecbfbc3b32ae6844cf65a6e9685b4edbe/pytato/distributed.py#L1279-L1280).

**Review by inducer — COMMENTED** [review]
Some thoughts below.

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

File: pytato/distributed.py [hunk 4]
```diff
@@ -1105,6 +1112,135 @@ class MissingRecvError(DistributedPartitionVerificationError):
...
+
+        self.input_name_to_recv_node: Dict[str, DistributedRecv] = {}
+        self.output_name_to_send_node: Dict[str, _SummarizedDistributedSend] = {}
+
+    # type-ignore-reason: dropped the extra `*args, **kwargs`.
+    def get_cache_key(self, expr: ArrayOrNames) -> int:  # type: ignore[override]
+        return id(expr)
+
+    def map_distributed_recv(  # type: ignore[override]
+            self, expr: DistributedRecv) -> None:
```
Comment (matthiasdiener) [inline pytato/distributed.py:1132]: For some reason this is never called...

File: test/test_distributed.py [hunk 5]
```diff
@@ -413,6 +441,13 @@ def _do_verify_distributed_partition(ctx_factory):
             stapled_to=recv)
 
     outputs = pt.make_dict_of_named_arrays({"out": send+send2})
+
+    if rank == 0:
+        with pytest.raises(MissingSendError):
+            pt.verify_distributed_dag_pre_partition(comm, outputs)
+    else:
+        pt.verify_distributed_dag_pre_partition(comm, outputs)
```
Comment (inducer) [inline test/test_distributed.py:449]: Find a way to make these less wordy.

**Comment by inducer** [comment]
Superseded by #404?

**Comment by inducer** [comment]
Superseded by #434 .

**Pull request closed without merging**